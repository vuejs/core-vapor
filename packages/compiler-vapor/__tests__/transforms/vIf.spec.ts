import { makeCompile } from './_utils'
import {
  IRNodeTypes,
  type IfIRNode,
  transformElement,
  transformInterpolation,
  transformOnce,
  transformVIf,
  transformVText,
} from '../../src'
import { NodeTypes } from '@vue/compiler-core'

const compileWithVIf = makeCompile({
  nodeTransforms: [
    transformOnce,
    transformInterpolation,
    transformVIf,
    transformElement,
  ],
  directiveTransforms: {
    text: transformVText,
  },
})

describe('compiler: v-if', () => {
  test('basic v-if', () => {
    const { code, vaporHelpers, ir, helpers } = compileWithVIf(
      `<div v-if="ok">{{msg}}</div>`,
    )

    expect(vaporHelpers).contains('createIf')
    expect(helpers.size).toBe(0)

    expect(ir.template).lengthOf(2)
    expect(ir.template).toMatchObject([
      {
        template: '<div></div>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        type: IRNodeTypes.FRAGMENT_FACTORY,
      },
    ])
    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.IF,
        id: 1,
        condition: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'ok',
          isStatic: false,
        },
        positive: {
          type: IRNodeTypes.BLOCK_FUNCTION,
          templateIndex: 0,
        },
      },
      {
        type: IRNodeTypes.APPEND_NODE,
        elements: [1],
        parent: 0,
      },
    ])

    expect(ir.dynamic).toMatchObject({
      id: 0,
      children: { 0: { id: 1 } },
    })

    expect(ir.effect).toEqual([])
    expect((ir.operation[0] as IfIRNode).positive.effect).lengthOf(1)

    expect(code).matchSnapshot()
  })

  test('template v-if', () => {
    const { code, ir } = compileWithVIf(
      `<template v-if="ok"><div/>hello<p v-text="msg"/></template>`,
    )
    expect(code).matchSnapshot()

    expect(ir.template).lengthOf(2)
    expect(ir.template[0]).toMatchObject({
      template: '<div></div>hello<p></p>',
      type: IRNodeTypes.TEMPLATE_FACTORY,
    })

    expect(ir.effect).toEqual([])
    expect((ir.operation[0] as IfIRNode).positive.effect).toMatchObject([
      {
        operations: [
          {
            type: IRNodeTypes.SET_TEXT,
            element: 3,
            value: {
              content: 'msg',
              type: NodeTypes.SIMPLE_EXPRESSION,
              isStatic: false,
            },
          },
        ],
      },
    ])
    expect((ir.operation[0] as IfIRNode).positive.dynamic).toMatchObject({
      id: 2,
      children: { 2: { id: 3 } },
    })
  })

  test('dedupe same template', () => {
    const { code, ir } = compileWithVIf(
      `<div v-if="ok">hello</div><div v-if="ok">hello</div>`,
    )
    expect(code).matchSnapshot()
    expect(ir.template).lengthOf(2)
  })

  test.todo('v-if with v-once')
  test.todo('component v-if')

  test('v-if + v-else', () => {
    const { code, ir, vaporHelpers, helpers } = compileWithVIf(
      `<div v-if="ok"/><p v-else/>`,
    )
    expect(code).matchSnapshot()
    expect(ir.template).lengthOf(3)
    expect(ir.template).toMatchObject([
      {
        template: '<div></div>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        template: '<p></p>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        type: IRNodeTypes.FRAGMENT_FACTORY,
      },
    ])

    expect(vaporHelpers).contains('createIf')
    expect(ir.effect).lengthOf(0)
    expect(helpers).lengthOf(0)
    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.IF,
        id: 1,
        condition: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'ok',
          isStatic: false,
        },
        positive: {
          type: IRNodeTypes.BLOCK_FUNCTION,
          templateIndex: 0,
        },
        negative: {
          type: IRNodeTypes.BLOCK_FUNCTION,
          templateIndex: 1,
        },
      },
      {
        type: IRNodeTypes.PREPEND_NODE,
        elements: [1],
        parent: 0,
      },
    ])
  })

  test('v-if + v-else-if', () => {
    const { code, ir } = compileWithVIf(
      `<div v-if="ok"/><p v-else-if="orNot"/>`,
    )
    expect(code).matchSnapshot()
    expect(ir.template).lengthOf(3)
    expect(ir.template).toMatchObject([
      {
        template: '<div></div>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        template: '<p></p>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      { type: IRNodeTypes.FRAGMENT_FACTORY },
    ])

    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.IF,
        id: 1,
        condition: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'ok',
          isStatic: false,
        },
        positive: {
          type: IRNodeTypes.BLOCK_FUNCTION,
          templateIndex: 0,
        },
        negative: {
          type: IRNodeTypes.IF,
          condition: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'orNot',
            isStatic: false,
          },
          positive: {
            type: IRNodeTypes.BLOCK_FUNCTION,
            templateIndex: 1,
          },
        },
      },
      {
        type: IRNodeTypes.PREPEND_NODE,
        elements: [1],
        parent: 0,
      },
    ])
  })

  test('v-if + v-else-if + v-else', () => {
    const { code, ir } = compileWithVIf(
      `<div v-if="ok"/><p v-else-if="orNot"/><template v-else>fine</template>`,
    )
    expect(code).matchSnapshot()
    expect(ir.template).lengthOf(4)
    expect(ir.template).toMatchObject([
      {
        template: '<div></div>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        template: '<p></p>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        template: 'fine',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      { type: IRNodeTypes.FRAGMENT_FACTORY },
    ])

    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.IF,
        id: 1,
        positive: {
          type: IRNodeTypes.BLOCK_FUNCTION,
          templateIndex: 0,
        },
        negative: {
          type: IRNodeTypes.IF,
          positive: {
            type: IRNodeTypes.BLOCK_FUNCTION,
            templateIndex: 1,
          },
          negative: {
            type: IRNodeTypes.BLOCK_FUNCTION,
            templateIndex: 2,
          },
        },
      },
      {
        type: IRNodeTypes.PREPEND_NODE,
        elements: [1],
        parent: 0,
      },
    ])
  })

  test('comment between branches', () => {
    const { code, ir } = compileWithVIf(`
      <div v-if="ok"/>
      <!--foo-->
      <p v-else-if="orNot"/>
      <!--bar-->
      <template v-else>fine</template>
    `)
    expect(code).matchSnapshot()
    expect(ir.template).lengthOf(4)
    expect(ir.template).toMatchObject([
      {
        template: '<div></div>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        template: '<!--foo--><p></p>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        template: '<!--bar-->fine',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      { type: IRNodeTypes.FRAGMENT_FACTORY },
    ])
  })

  describe.todo('errors')
  describe.todo('codegen')
  test.todo('v-on with v-if')
})
