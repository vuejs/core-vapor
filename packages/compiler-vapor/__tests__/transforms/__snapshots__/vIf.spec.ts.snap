// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: v-if > basic v-if 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, children as _children, renderEffect as _renderEffect, setText as _setText, append as _append } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const t1 = _fragment()
  const n0 = t1()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    const { 0: [n3],} = _children(n2)
    _renderEffect(() => {
      _setText(n3, _ctx.msg)
    })
    return n2
  })
  _append(n0, n1)
  return n0
}"
`;

exports[`compiler: v-if > comment between branches 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, prepend as _prepend } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const t1 = _template("<!--foo--><p></p>")
  const t2 = _template("<!--bar-->fine")
  const t3 = _fragment()
  const n0 = t3()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (_ctx.orNot), () => {
    const n3 = t1()
    return n3
  }, () => {
    const n4 = t2()
    return n4
  }))
  _prepend(n0, n1)
  return n0
}"
`;

exports[`compiler: v-if > dedupe same template 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, append as _append } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div>hello</div>")
  const t1 = _fragment()
  const n0 = t1()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    return n2
  })
  const n3 = _createIf(() => (_ctx.ok), () => {
    const n4 = t0()
    return n4
  })
  _append(n0, n1, n3)
  return n0
}"
`;

exports[`compiler: v-if > template v-if 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, children as _children, renderEffect as _renderEffect, setText as _setText, append as _append } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>hello<p></p>")
  const t1 = _fragment()
  const n0 = t1()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    const { 2: [n3],} = _children(n2)
    _renderEffect(() => {
      _setText(n3, _ctx.msg)
    })
    return n2
  })
  _append(n0, n1)
  return n0
}"
`;

exports[`compiler: v-if > v-if + v-else 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, prepend as _prepend } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const t1 = _template("<p></p>")
  const t2 = _fragment()
  const n0 = t2()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    return n2
  }, () => {
    const n3 = t1()
    return n3
  })
  _prepend(n0, n1)
  return n0
}"
`;

exports[`compiler: v-if > v-if + v-else-if + v-else 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, prepend as _prepend } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const t1 = _template("<p></p>")
  const t2 = _template("fine")
  const t3 = _fragment()
  const n0 = t3()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (_ctx.orNot), () => {
    const n3 = t1()
    return n3
  }, () => {
    const n4 = t2()
    return n4
  }))
  _prepend(n0, n1)
  return n0
}"
`;

exports[`compiler: v-if > v-if + v-else-if 1`] = `
"import { template as _template, fragment as _fragment, createIf as _createIf, prepend as _prepend } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const t1 = _template("<p></p>")
  const t2 = _fragment()
  const n0 = t2()
  const n1 = _createIf(() => (_ctx.ok), () => {
    const n2 = t0()
    return n2
  }, () => _createIf(() => (_ctx.orNot), () => {
    const n3 = t1()
    return n3
  }))
  _prepend(n0, n1)
  return n0
}"
`;
